aggregators/aggregator.go

```diff
@@ -48,86 +40,21 @@ var (
 // same processing time bucket and thereafter the processing time
 // bucket is advanced in factors of aggregation interval.
 type Aggregator struct {
-	db  *pebble.DB
-	cfg Config
+	db           *pebble.DB
+	writeOptions *pebble.WriteOptions
+	cfg          Config
 
-	mu             sync.Mutex
+	mu             sync.RWMutex
+	writers        sync.Map
```

Comment:
It seems that `writers` is always accessed within `Aggregator#mu#Lock`. Maybe we don't need the `sync.Map`? I think we should also be good with a slice.

Reply:
That mutex isn't necessarily held when closing a writer.

Reply:
a slice requires O(N) writer removal though

Reply:
maybe just a `map[*Writer]struct{}`?

Reply:
(oh I'm late to the party, ignore my comments. Thanks for @axw 's explanation. I see that Writer.Close only holds w.mu.)

Reply:
> That mutex isn't necessarily held when closing a writer.

Sorry, didn't get this. I think if we were to use a slice, for example `[]*Writer` then we would only need to consider locking the slice when we add to the slice or remove from the slice or iterate over the slice. From what I can see, we take the lock in all the cases, am I missing something?

> a slice requires O(N) writer removal though

We remove all at one time, we can just iterate over the slice from the end and mark each of them nil. No need for O(N) I think.

Reply:
> I see that Writer.Close only holds w.mu.)

Yes but the caller to `Writer.Close` i.e. `Aggregator.Close` gets a lock. I was suggesting to keep the access of `writers` to only the aggregator code and remove the requirement of sync based structure.

Reply:
> Yes but the caller to Writer.Close i.e. Aggregator.Close gets a lock.

I believe user code will call `Writer.Close` as well, not only `Aggregator.Close`.

Reply:
> I believe user code will call Writer.Close as well, not only Aggregator.Close

Right. At least in theory, you could have multiple independent writers and they can both independently close their writers. For example, see BenchmarkAggregateBatch.

Reply:
> I believe user code will call Writer.Close as well, not only Aggregator.Close.

Ah, good point, didn't realize this.